### PR TITLE
Bug 217681: compiler should not crash on public+private metadata

### PR DESCRIPTION
**Pull request by jcouv**
This fixes this [Watson-reported crash](https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?id=217681&triage=true&fullScreen=true&_a=edit) for update 3.

The crash occurs when accessibility flags on a method are set to 7 (which is all the flags, which means both public and private).

This problem existed on the VB side for a long time. On the C# side, before [this change](https://github.com/dotnet/roslyn/commit/defaa1b691b34e7338576afabaad60eb61a3b4ea), those switch statements handled the default case with a debug assert and defaulting to "private".

**On the C# side:**
- native compiler treated those flags as "private" and reports errors:
```
consumer.cs(5,15): error CS1061: 'C' does not contain a definition for 'M' and no extension method 'M' accepting a first argument of type 'C' could be found (are you missing a using directive or an assembly reference?)
consumer.cs(6,40): error CS1061: 'C' does not contain a definition for 'F' and no extension method 'F' accepting a first argument of type 'C' could be found (are you missing a using directive or an assembly reference?)
consumer.cs(7,13): error CS0426: The type name 'C2' does not exist in the type 'C'
```
- compiler 1.2 compiler failed silently (no executable produced)
- future branch crashes
- after fix, errors are reported:
```
consumer.cs(5,15): error CS0122: 'C.M()' is inaccessible due to its protection level
consumer.cs(6,40): error CS0122: 'C.F' is inaccessible due to its protection level
consumer.cs(7,13): error CS0426: The type name 'C2' does not exist in the type 'C'
```
  
  

**On the VB side:**
- native compiler successfully compiles the method and field case (those only yield runtime error `System.TypeLoadException: Invalid Field Access Flags`) and reported an accessibility error on the nested type.
- compiler 1.2 generated errors:
```
error BC30390: 'C.Private Overloads Sub M()' is not accessible in this context because it is 'Private'.
error BC30389: 'C.F' is not accessible in this context because it is 'Private'.
error BC30389: 'C.C2' is not accessible in this context because it is 'Protected Friend'.
error BC30390: 'C2.Private Overloads Sub M2()' is not accessible in this context because it is 'Private'.
```
- future branch crashes
- after fix, the same errors as 1.2 are generated again. I will notify the compat council of this breaking with the native compiler.

### CODE REVIEW

**Comment by jcouv** [comment]
CC @dotnet/roslyn-compiler for review.

**Comment by jcouv** [comment]
@MattGertz Can you approve for update 3?
The compat council approved the small compat break.

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler Please review for update 3.

**Comment by MattGertz** [comment]
Approved pending CR signoffs.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingMethods.cs [hunk 1]
```diff
@@ -1346,5 +1346,38 @@ public void Repro666162()
...
+        {
+            string source =
+@"
+public class D
+{
+   public static void Main()
+   {
+      new C().M();
+      System.Console.WriteLine(new C().F);
+      new C.C2().M2();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Metadata/PE/LoadingMethods.cs]: I assume ```()``` after C2 are not intentional.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs [hunk 2]
```diff
@@ -413,7 +413,8 @@ public override Accessibility DeclaredAccessibility
                         break;
 
                     default:
-                        throw ExceptionUtilities.UnexpectedValue(_flags & FieldAttributes.FieldAccessMask);
+                        access = Accessibility.Private;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: This bug was meant to cover `public + private` accessibility flags.  This change expands it to all forms of accessibility not caught above.  Is that intentional?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: There are only eight values possible

---
In reply to: [64623823](https://github.com/dotnet/roslyn/pull/11547#discussion_r64623823) [](ancestors = 64623823)

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb [hunk 3]
```diff
@@ -925,5 +925,35 @@ BC30657: 'VT' has a return type that is not supported or parameter types that ar
...
+      test2.M2()
+   End Sub
+End Class
+    </file>
+         </compilation>
+
+            Dim references = {MetadataReference.CreateFromImage(TestResources.SymbolsTests.Metadata.PublicAndPrivateFlags)}
+
+            Dim comp = CreateCompilationWithMscorlib(source, references:=references)
+            comp.VerifyDiagnostics(
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb]: Please use AssertTheseDiagnostics helper instead.

**Comment by AlekseyTs** [comment]
:shipit:

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs (hunk 2 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: See the mask at the top of the switch case. The default case only covered this one remaining value.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb (hunk 3 above)
Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb]: What is the advantage of this method over the current approach?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEFieldSymbol.cs]: Gotcha.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb (hunk 3 above)
Comment (jcouv) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb]: Done

Comment (jcouv) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb]: @jaredpar It captures more details about the diagnostics and seems more readable.

Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb]: @jaredpar This is the preferred way to baseline diagnostics for VB (human readable with squiggle). See the previous unit-test, for example.

Comment (jaredpar) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/Metadata/PE/LoadingMethods.vb]: Gotcha.

**Comment by jaredpar** [comment]
👍

**Pull request merged**